File: spindare-ios/SpindareKit/Sources/SpindareKit/Services/AppEnvironment.swift
import Foundation
import ClerkKit

/// The app's composition root, reduced to a set of swappable service slots.
///
/// Every view and view-model injects its service via a default argument. Those
/// defaults used to be `= MockFeedService()` literally; they're now
/// `= AppEnvironment.feedService`, so the one place that decides Mock-vs-Live is
/// here. `SpindareApp.init()` overwrites these with `Live*Service` values backed
/// by a real `APIClient` *before* the first view renders — Swift re-evaluates a
/// default-argument expression at each call site, so whatever's assigned here at
/// launch is what every subsequently-constructed view picks up.
///
/// Chat is live now too. Speedys and Zone still have no `Live` implementation,
/// so they keep their own `Mock*Service()` defaults and are untouched by this.
///
/// `nonisolated(unsafe)`: these are written exactly once, synchronously, during
/// app launch before any concurrency exists, then only ever read. The stored
/// values are all `Sendable` (the protocols require it). The single-writer-at-
/// startup shape is what makes the opt-out honest rather than a race waiting to
/// happen.
public enum AppEnvironment {
    nonisolated(unsafe) public static var feedService: any FeedServing = MockFeedService()
    nonisolated(unsafe) public static var profileService: any ProfileServing = MockProfileService()
    nonisolated(unsafe) public static var socialService: any SocialServing = MockSocialService()
    nonisolated(unsafe) public static var notificationService: any NotificationServing = MockNotificationService()
    nonisolated(unsafe) public static var searchService: any SearchServing = MockSearchService()
    nonisolated(unsafe) public static var chatService: any ChatServing = MockChatService()

    /// Uploads media to R2 via the Nest storage endpoints. `nil` in mock mode:
    /// call sites fall back to writing a local file so posting still works
    /// entirely offline. Set to a real uploader only when running live.
    nonisolated(unsafe) public static var mediaUploader: MediaUploader?

    /// Assigns the live service set. Called from `bootstrap` when a backend
    /// URL is configured.
    public static func useLive(api: APIClient) {
        feedService = LiveFeedService(api: api)
        profileService = LiveProfileService(api: api)
        socialService = LiveSocialService(api: api)
        notificationService = LiveNotificationService(api: api)
        searchService = LiveSearchService(api: api)
        chatService = LiveChatService(api: api)
        mediaUploader = MediaUploader(api: api)
    }

    /// The one call the app's `@main` entry point makes at launch.
    ///
    /// Configures Clerk (always — auth is real regardless of backend state),
    /// then, *only if* a backend URL is set, swaps every service onto the live
    /// backend and wires the API client to mint a fresh Clerk token per request.
    /// With no URL the app keeps serving on-device mock data, so it stays fully
    /// usable before the backend is deployed. Kept here, rather than in the app
    /// target, so `ClerkKit` stays an implementation detail of this package and
    /// the app target needn't depend on it directly.
    /// UserDefaults key for the QA "use on-device mock data" switch. Shared with
    /// the `@AppStorage` toggle in Settings.
    public static let testDataDefaultsKey = "useTestData"

    @MainActor
    public static func bootstrap(clerkPublishableKey: String, apiBaseURL: String) {
        Clerk.configure(publishableKey: clerkPublishableKey)

        // QA override: stay entirely on-device mock data even with a backend
        // configured. Read once at launch (services are single-writer-at-startup
        // by design), so flipping it in Settings applies on the next launch —
        // the toggle says as much. Lets the team keep testing on demo data
        // through the pre-launch window without touching prod.
        if UserDefaults.standard.bool(forKey: testDataDefaultsKey) { return }

        let trimmed = apiBaseURL.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return }

        let api = APIClient(baseURL: trimmed)
        useLive(api: api)
        Task {
            await api.setTokenProvider {
                try? await Clerk.shared.auth.getToken()
            }
        }
    }

    /// Ends the Clerk session. Kept here so `AppRouter`/`SettingsView` can log
    /// out without importing ClerkKit — the session survived a plain
    /// `didSignOut()` before, which is why "log out" didn't actually log you out.
    @MainActor
    public static func signOut() async {
        try? await Clerk.shared.auth.signOut()
    }

    /// The identity restored from a persisted Clerk session at launch, if any.
    public struct RestoredIdentity: Sendable {
        public let userId: String
        /// Optional on purpose: nil means "we don't know it yet", which the UI
        /// renders as a placeholder. It must never be a placeholder *value*, or
        /// it can be persisted as the user's real handle.
        public let username: String?
        public let email: String?
        public let avatarURL: String?
    }

    /// Loads any persisted Clerk session at launch and, if one exists, resolves
    /// the signed-in identity (pulling the real username/avatar from the backend
    /// profile). Returns nil when there's no session — the app then shows
    /// onboarding as before. Without this, a returning user was forced back
    /// through sign-in on every launch even though Clerk still held their session.
    @MainActor
    public static func restoreSession() async -> RestoredIdentity? {
        // ClerkKit 1.3 restores the persisted session by fetching the client
        // (which carries the current session and user) — the older one-shot
        // `Clerk.load()` was split into refreshClient()/refreshEnvironment().
        try? await Clerk.shared.refreshClient()
        guard let user = Clerk.shared.user else { return nil }

        let profile = (try? await profileService.currentProfile()) ?? nil

        // Never substitute a placeholder here. This value becomes
        // `router.username`, which Settings seeds its text field from and then
        // *saves* — so a display fallback like "you" leaking in was how a real
        // account ended up renamed to "you", with the email and avatar blanked
        // alongside it. If the backend profile can't be read, fall back to what
        // Clerk itself knows and otherwise leave it nil, so the UI shows a
        // placeholder without any of it being mistaken for real data.
        return RestoredIdentity(
            userId: user.id,
            username: profile?.username ?? user.username,
            email: profile?.email ?? user.primaryEmailAddress?.emailAddress,
            avatarURL: profile?.photoURL ?? (user.hasImage ? user.imageUrl : nil)
        )
    }
}

File: spindare-android/app/src/main/java/al/spind/spindare/services/AppEnvironment.kt
package al.spind.spindare.services

import android.content.Context
import com.clerk.api.Clerk
import com.clerk.api.network.serialization.ClerkResult
import al.spind.spindare.net.ApiClient
import al.spind.spindare.net.SpindareApi

/**
 * The composition switchboard, ported from SpindareKit/Services/AppEnvironment.swift.
 *
 * Every screen reads its dependencies from here, so flipping the whole app
 * between mock and live is one call at startup rather than an edit in each
 * view. Written once at launch and read everywhere after — single-writer by
 * design, which is why plain `var`s are safe without synchronisation.
 */
object AppEnvironment {

    /** Set only while a backend is configured; null keeps the app on mock data. */
    var api: SpindareApi? = null
        private set

    var feed: FeedServing = MockFeedService()
        private set

    var profile: ProfileServing = MockProfileService()
        private set

    var social: SocialServing = MockSocialService()
        private set

    var notification: NotificationServing = MockNotificationService()
        private set

    var search: SearchServing = MockSearchService()
        private set

    var chat: ChatServing = MockChatService()
        private set

    var speedy: SpeedyServing = MockSpeedyService()
        private set

    var zone: ZoneServing = MockZoneService()
        private set

    var mediaUploader: MediaUploader? = null
        private set

    /** True once [bootstrap] has pointed the services at a real backend. */
    val isLive: Boolean get() = api != null

    /**
     * @param apiBaseUrl empty keeps the app on on-device mock data while auth
     *   stays real Clerk — the same escape hatch the iOS build has, and what
     *   makes the app demoable with no network.
     */
    fun bootstrap(context: Context, clerkPublishableKey: String, apiBaseUrl: String, debug: Boolean) {
        Clerk.initialize(context, publishableKey = clerkPublishableKey)

        val trimmed = apiBaseUrl.trim()
        if (trimmed.isEmpty()) return

        val client = ApiClient.create(
            baseUrl = trimmed,
            debug = debug,
            // Per-request, so an expired token is re-fetched rather than
            // 401-ing every call until the app restarts.
            tokenProvider = { (Clerk.auth.getToken() as? ClerkResult.Success)?.value },
        )
        api = client
        useLive(client)
    }

    private fun useLive(api: SpindareApi) {
        feed = LiveFeedService(api)
        profile = LiveProfileService(api)
        social = LiveSocialService(api)
        notification = LiveNotificationService(api)
        search = LiveSearchService(api)
        chat = LiveChatService(api)
        speedy = LiveSpeedyService()
        zone = LiveZoneService()
        mediaUploader = MediaUploader(api)
    }

    /** Resets all mock data to its initial seed state. No-op in live mode. */
    fun resetMocks() {
        if (isLive) return
        feed = MockFeedService()
        profile = MockProfileService()
        social = MockSocialService()
        notification = MockNotificationService()
        search = MockSearchService()
        chat = MockChatService()
        speedy = MockSpeedyService()
        zone = MockZoneService()
    }

    /** Ends the Clerk session. Kept here so UI never imports the Clerk SDK directly. */
    suspend fun signOut() {
        Clerk.auth.signOut()
    }

    /**
     * The identity restored from a persisted Clerk session at launch, if any.
     *
     * [username] is nullable on purpose: null means "we don't know it yet",
     * which the UI renders as a placeholder. It must never be a placeholder
     * *value* — on iOS a literal `"you"` fallback leaked all the way into the
     * database as people's real handle. Same trap exists here.
     */
    data class RestoredIdentity(
        val userId: String,
        val username: String?,
        val email: String?,
        val avatarUrl: String?,
    )

    // `hasImage` is marked deprecated in clerk-android 1.0.36 with a message
    // pointing at `hasUploadedImage` — a property that does not exist in this
    // version. Using the deprecated one is the only option until Clerk actually
    // ships the replacement; revisit on the next SDK bump.
    @Suppress("DEPRECATION")
    fun currentIdentity(): RestoredIdentity? {
        val user = Clerk.user ?: return null
        return RestoredIdentity(
            userId = user.id,
            username = user.username,
            email = user.primaryEmailAddress?.emailAddress,
            // imageUrl is non-null and always populated — Clerk serves a
            // generated avatar when none was uploaded — so `hasImage` is what
            // distinguishes a real picture from a placeholder.
            avatarUrl = user.imageUrl.takeIf { it.isNotEmpty() && user.hasImage },
        )
    }
}
